---
title: "ensembl_cardiac_diseases_panel_20200715"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressMessages(library(knitr))
suppressMessages(library(data.table))
suppressMessages(library(tidyverse))

outlier <- function (x) {
  IQR <- quantile(x)[[4]] - quantile(x)[[2]]
  low <- quantile(x)[[2]] - 1.5*IQR
  high <- quantile(x)[[4]] + 1.5*IQR
  out_val <- x[x < low | x > high]
  return(out_val)
}
```

```{r set p treshold, echo=FALSE}
Pt <- 0.05
```


```{r sample info, echo=FALSE}
pheno <- read.table('input/plink_data/pheno.txt',header = T)
pheno %>% na.omit %>% ggplot(aes(factor(Y1))) + geom_bar(fill=c('#27384A','#48C095'),width=0.5) +
  geom_text(stat='count', aes(label=..count..), vjust=-1) + 
  theme_classic() + 
  ylim(c(0,1000)) + 
  ylab('Number of individuals') +
  xlab('Covid severity') +
  scale_x_discrete(labels=c('Control','Severe')) + 
  ggtitle(paste('Distribution of covid severity in cohort of',pheno %>% 
                  na.omit() %>% nrow(), 'individuals',sep = ' ')) +
  theme(plot.title = element_text(hjust = 0.5))
```



```{r af plot, echo=FALSE}
cardiac_fish <- fread('output/cardiac_panel.assoc.fisher') %>% na.omit()
cardiac_sig <- cardiac_fish %>%
  filter(P < 0.005)

vep_pos <- cardiac_sig %>% select(CHR,BP)
vep_pos$POS <- paste('chr',vep_pos$CHR,sep='')
vep_pos$POS <- paste(vep_pos$POS, vep_pos$BP,sep=':')
write.table(vep_pos$POS,'input/cardiac_pos.tsv',quote = F,
             col.names = F,
             row.names = F,
             sep='\t')

cardiac_sig %>%
  ggplot(aes(x=F_U,y=F_A)) + geom_point(col='#48C095') +
  geom_smooth(method = 'lm', col = '#27384A', formula = 'y~x',se=F) +
  ylab('Frequency of allele in cases') +
  xlab('Frequency of allele in controls') +
  ggtitle('Frequency of alleles in cases and controls', 
          subtitle = paste('Fisher exact test P value <', Pt,sep = ' ')) + 
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),plot.subtitle = element_text(hjust=0.5))
  

```

```{r OR plot, echo=FALSE}
cardiac_fish %>% 
  filter(P < Pt) %>%
  ggplot(aes(x=OR,y=1)) + 
  geom_violin(fill='#48C095', col='#27384A') +
  geom_boxplot(width=0.1) +
  ylab('') +
  xlab('Odds ratio') +
  ggtitle('Estimated odds ratio of minor allele', 
          subtitle = paste('Fisher exact test P value <', Pt,sep = ' ')) + 
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),plot.subtitle = element_text(hjust=0.5)) +
  theme(axis.text.y = element_blank())
  
```


```{r OR outliers, echo=FALSE}
or_out <- cardiac_fish %>% filter(P < Pt) %>% filter(OR %in% outlier(cardiac_fish$OR))

```

```{r, echo =FALSE}
vep <- read.csv('input/cardiac_panel.vep',skip = 40,header=T,sep='\t')
high_impact <- vep %>% select(X.Uploaded_variation,Location,Allele,Gene,IMPACT,Consequence)

high_sig <- cardiac_sig %>% left_join(high_impact, by=c('SNP'='X.Uploaded_variation'))

sig_impact <- high_sig %>% filter(IMPACT=='LOW' | IMPACT=='MODERATE')
sig_conseq <- high_sig %>% filter(Consequence != 'synonymous_variant' & Consequence != 'intron_variant')

kable(
  sig_impact %>% select(SNP,Location,Allele,Gene,F_A,F_U,P,OR,IMPACT,Consequence)
)
```


