---
title: "OAS_panel"
subtitle: 'Results on cohort of 943 unrelated individuals'
output: 
  pdf_document: default
  github_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressMessages(library(knitr))
suppressMessages(library(data.table))
suppressMessages(library(tidyverse))
suppressMessages(library(qvalue))

outlier <- function (x) {
  IQR <- quantile(x)[[4]] - quantile(x)[[2]]
  low <- quantile(x)[[2]] - 1.5*IQR
  high <- quantile(x)[[4]] + 1.5*IQR
  out_val <- x[x < low | x > high]
  return(out_val)
}
```

## 1. Cohort information

```{r sample info, echo=FALSE}
prepheno <- read.table('../input/prepheno.csv',header=T,sep='\t')
to_include <- read.table('../../../genom_polaka/SAMPLES_TO_INCLUDE_210519.txt',sep='\t')

pheno <- prepheno %>% filter(prepheno$Individual_ID %in% to_include$V1)
pheno$Covid_severity <- ifelse(pheno$Covid_severity == 'COV_SEVERE', 1,0) 
  pheno%>%
  ggplot(aes(factor(Covid_severity))) + geom_bar(fill=c('#27384A','#48C095'),width=0.5) +
  geom_text(stat='count', aes(label=..count..), vjust=-1) + 
  ylim(c(0,1000)) + 
  ylab('Number of individuals') +
  xlab('Covid severity') +
  scale_x_discrete(labels=c('Control','Severe')) + 
  ggtitle(paste('Distribution of covid severity in cohort of',pheno %>% nrow(), 'individuals',sep = ' ')) +
  theme_classic() +  
  theme(plot.title = element_text(hjust = 0.5))
```

Control label include mild benign and resistant response to COVID-19 infection

\newpage
## 2. Results

Association study was performed by comparison of allele counts in cases 
and controls with Chi-square test. Significance of variants is expressed in form of q-values 
which are FDR adjusted p-values.


```{r af plot, echo=FALSE}
group.colors <- c(significant = "#BC0020", `not significant` ="#48C095")

oas_fish <- fread('../output/OAS.assoc')  %>% distinct()
qobj <- qvalue(p = oas_fish$P)
oas_fish <- oas_fish %>% mutate(q = qobj$qvalues) %>% arrange(q) 
oas_fish$significance <- ifelse(oas_fish$q < 0.32 ,'significant','not significant') 

vep_pos <- oas_fish %>% select(CHR,BP)
vep_pos$POS <- paste('chr',vep_pos$CHR,sep='')
vep_pos$POS <- paste(vep_pos$POS, vep_pos$BP,sep=':')
write.table(vep_pos$POS,'../input/oas_pos.tsv',quote = F,
             col.names = F,
             row.names = F,
             sep='\t')

oas_fish %>% na.omit() %>%
  ggplot(aes(x=C_U,y=C_A,col=significance)) +
  geom_point() +
  ylab('Counts of allele in cases') +
  xlab('Counts of allele in controls') +
  ggtitle('Minor allele count in cases and controls') +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust=0.5)) +
  scale_color_manual(values=group.colors)


```

### Manhattan plot 
Red points are SNPs with qvalue < 0.32

```{r q_manhattan_plot, echo=F}
oas_fish$chr_col <- 'chr'
#oas_fish$chr_col <- ifelse(oas_fish$CHR %% 2 == 0, 'even','odd')
oas_fish$chr_col <- ifelse(oas_fish$q < 0.32, 's',oas_fish$chr_col)

chr.group <- c(chr = '#48C095',s='#BC0020')

oas_fish %>% 
  filter(is.na(P) == F) %>%
  ggplot(aes(x=BP/1e+06,y=-log10(P))) +
  geom_jitter(aes(col=chr_col)) +
  scale_color_manual(values = chr.group) +
  theme_classic() +
  theme(legend.position = 'none') +
  xlab('Position (mbp)')

```




```{r, echo =FALSE}
vep <- read.csv('../input/oas.vep.tsv',header=T,sep='\t')

vep_oas <- oas_fish %>%
  filter(significance=='significant') %>%
  select(SNP) %>% left_join(vep, by = c('SNP'='Uploaded_variation')) %>%
  right_join(oas_fish, by = 'SNP')  %>%
  filter(is.na(SYMBOL) == F) %>%
  select(SNP,Existing_variation,SYMBOL,A1,A2, CLIN_SIG, q,P,C_A,C_U,OR,L95,U95,significance) %>% 
  rename('qvalue' = q, 'case_count' = C_A, 'ctrl_count'=C_U) %>% 
  arrange(qvalue)

write.table(vep_oas,'../output/OAS_variants.tsv',sep='\t',col.names = T,row.names = F,quote = F)

```



```{r OR plot, echo=FALSE}
orplot <- vep_oas %>%
  filter(is.na(OR) == F & is.na(Existing_variation) == F) %>% distinct()

orplot %>%
  ggplot(aes(y=reorder(SNP,-OR), x=OR)) +
  geom_point(col="#48C095") +
  geom_text(aes(label=SYMBOL),hjust=-0.2, vjust=0.2) +
  geom_vline(xintercept = 1,linetype='dashed',col="#BC0020") +
  ylab('SNP') +
  xlab('Odds ratio') +
  xlim(c(0,10)) +
  ggtitle('Estimated odds ratio (OR) of minor allele occurence',
          subtitle = 'Red dashed line: OR = 1.0') +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust=0.5))


```

\newpage

```{r, echo=FALSE}
kable(vep_oas %>% 
        select(SNP,SYMBOL,
               A2,A1,case_count,ctrl_count,qvalue,OR) %>%
        arrange(qvalue),
        caption = "Variants with q-value < 0.32"
        )

```

