---
title: "OAS_panel"
output: 
  github_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressMessages(library(knitr))
suppressMessages(library(data.table))
suppressMessages(library(tidyverse))
suppressMessages(library(qvalue))

outlier <- function (x) {
  IQR <- quantile(x)[[4]] - quantile(x)[[2]]
  low <- quantile(x)[[2]] - 1.5*IQR
  high <- quantile(x)[[4]] + 1.5*IQR
  out_val <- x[x < low | x > high]
  return(out_val)
}
```

## 1. Cohort information

```{r sample info, echo=FALSE}
prepheno <- read.table('../input/prepheno.csv',header=T,sep='\t')
to_include <- read.table('../../../genom_polaka/SAMPLES_TO_INCLUDE_210519.txt',sep='\t')

pheno <- prepheno %>% filter(prepheno$Individual_ID %in% to_include$V1)
pheno$Covid_severity <- ifelse(pheno$Covid_severity == 'COV_SEVERE', 1,0) 
  pheno%>%
  ggplot(aes(factor(Covid_severity))) + geom_bar(fill=c('#27384A','#48C095'),width=0.5) +
  geom_text(stat='count', aes(label=..count..), vjust=-1) + 
  ylim(c(0,1000)) + 
  ylab('Number of individuals') +
  xlab('Covid severity') +
  scale_x_discrete(labels=c('Control','Severe')) + 
  ggtitle(paste('Distribution of covid severity in cohort of',pheno %>% nrow(), 'individuals',sep = ' ')) +
  theme_classic() +  
  theme(plot.title = element_text(hjust = 0.5))
```

Control label include mild benign and resistant response to COVID-19 infection


<!-- ```{r PCA info, echo=FALSE} -->
<!-- pca <- read.table('output/cardiac.eigenvec') -->
<!-- pca_pheno <- pheno %>% select(Individual_ID,Covid_severity) %>% left_join(pca,by = c('Individual_ID'='V2')) -->

<!-- ``` -->

\newpage
## 2. Results

Association study was performed by comparison of allele counts in cases 
and controls with Chi-square test. Significance of variants is expressed in form of q-values 
which are FDR adjusted p-values.


```{r af plot, echo=FALSE}
group.colors <- c(significant = "#BC0020", `not significant` ="#48C095")

cardiac_fish <- fread('../output/OAS.assoc')  %>% distinct()
qobj <- qvalue(p = cardiac_fish$P)
cardiac_fish$q <- qobj$qvalues
cardiac_fish$significance <- ifelse(cardiac_fish$q < 0.32 ,'significant','not significant') 

vep_pos <- cardiac_fish %>% select(CHR,BP)
vep_pos$POS <- paste('chr',vep_pos$CHR,sep='')
vep_pos$POS <- paste(vep_pos$POS, vep_pos$BP,sep=':')
write.table(vep_pos$POS,'../input/oas_pos.tsv',quote = F,
             col.names = F,
             row.names = F,
             sep='\t')

cardiac_fish %>% na.omit() %>%
  ggplot(aes(x=F_U,y=F_A,col=significance)) +
  geom_point(alpha=0.5) +
  ylab('Frequency of allele in cases') +
  xlab('Frequency of allele in controls') +
  ggtitle('Minor allele frequency in cases and controls') +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust=0.5)) +
  scale_color_manual(values=group.colors)


```





```{r, echo =FALSE}
vep <- read.csv('../input/oas.vep.tsv',header=T,sep='\t')
vep_cardiac <- cardiac_fish %>% filter(significance=='significant') %>%
  select(SNP) %>% left_join(vep, by = c('SNP'='Uploaded_variation')) %>%
  right_join(cardiac_fish, by = 'SNP')  %>%
  select(SNP,Existing_variation,SYMBOL,Consequence,IMPACT, CLIN_SIG, q,F_A,F_U,OR) %>% 
  rename('qvalue' = q, 'case_frequency' = F_A, 'ctrl_frequency'=F_U, 
                       'minor_allele_OR' = OR) %>% 
  filter(grepl("rs",Existing_variation)) %>%
  arrange(desc(case_frequency))  

write.table(vep_cardiac,'../output/OAS_variants.tsv',sep='\t',col.names = T,row.names = F,quote = F)


```



```{r OR plot, echo=FALSE}
cardiac_fish %>% filter(significance == 'significant') %>%
  left_join(vep_cardiac, by = 'SNP') %>%  
  filter(is.na(minor_allele_OR) == F & is.na(Existing_variation) == F) %>%
  ggplot(aes(y=Existing_variation,x=minor_allele_OR)) +
  geom_point(col="#48C095") +
  geom_text(aes(label=SYMBOL),hjust=-0.2, vjust=0.2) +
  geom_vline(xintercept = 1,linetype='dashed',col="#BC0020") +
  ylab('rsid') +
  xlab('Odds ratio') +
  xlim(c(0,10)) +
  ggtitle('Estimated odds ratio (OR) of minor allele occurence',
          subtitle = 'Red dashed line: OR = 1.0') +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust=0.5))


```

\newpage

```{r, echo=FALSE}
kable(vep_cardiac %>% select(-SNP,-CLIN_SIG,-Consequence) %>%
        rename('rsid'=Existing_variation, 'OR'=minor_allele_OR),
        caption = "Variants with q-value < 0.32"
        )

```

