---
title: "ensembl_cardiac_diseases_panel_20200715"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressMessages(library(knitr))
suppressMessages(library(data.table))
suppressMessages(library(tidyverse))
suppressMessages(library(qvalue))

outlier <- function (x) {
  IQR <- quantile(x)[[4]] - quantile(x)[[2]]
  low <- quantile(x)[[2]] - 1.5*IQR
  high <- quantile(x)[[4]] + 1.5*IQR
  out_val <- x[x < low | x > high]
  return(out_val)
}
```

Results on 1076 unrelated individuals
8129 variants with MAF > 0.01 and LD pruning (indep-pairwise 50 5 0.05)


```{r loading_data, echo=F}
severe <-   fread('../../../output_data/MS_gwas_data/output/cardiac/cardiac_plink.assoc.logistic')
severe <- severe %>% 
  mutate(q = qvalue(P)$qvalue) %>% 
  na.omit() %>%
  arrange(q)
# 
# resistance <- fread('../output/CARDIAC/multisample_CARDIAC_filtered.pruned.resistance.assoc.logistic')
# resistance <- resistance %>% 
#   mutate(q = qvalue(P)$qvalue) %>% 
#   na.omit() %>%
#   arrange(q)
```

```{r echo=F}
kable(severe %>% filter(q <= min(severe$q)) %>%
        select(-c(CHR,BP,A1)) %>%
        mutate(SYMBOL = c('HADHA','HADHB')) %>%
        relocate(SYMBOL,.before = SNP)
        , caption = 'https://onlinelibrary.wiley.com/doi/10.1002/jmd2.12165')

```

```{r q_manhattan_plot, echo=F}
severe$chr_col <- NA
severe$chr_col <- ifelse(severe$CHR %% 2 == 0, 'even','odd')
severe$chr_col <- ifelse(severe$q <= min(severe$q), 's',severe$chr_col)

chr.group <- c(even = '#27384A',odd = '#48C095',s='#BC0020')

severe %>% 
  filter(is.na(P) == F) %>%
  ggplot(aes(x=CHR,y=-log10(P))) +
  geom_jitter(aes(col=chr_col)) +
  scale_color_manual(values = chr.group) +
  theme_classic() +
  theme(legend.position = 'none') +
  scale_x_continuous(labels = c(1:22,'X'),breaks = 1:23) +
  xlab('Chromosomes')
```