---
title: "ensembl_cardiac_diseases_panel_20200715"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressMessages(library(knitr))
suppressMessages(library(data.table))
suppressMessages(library(tidyverse))
suppressMessages(library(qvalue))

outlier <- function (x) {
  IQR <- quantile(x)[[4]] - quantile(x)[[2]]
  low <- quantile(x)[[2]] - 1.5*IQR
  high <- quantile(x)[[4]] + 1.5*IQR
  out_val <- x[x < low | x > high]
  return(out_val)
}
```

Results on 1076 unrelated individuals and 8,129 SNPs

```{r loading_data, echo=F}
vep <- fread('cardiac_fullset_sorted.vcf.gz.vep.tsv.gz',header=T) %>% rename(SNP='#Uploaded_variation') %>%
  select(SNP,Gene,SYMBOL,IMPACT,Consequence)


or  <- fread('/mnt/theta/output_data/MS_gwas_data/output/cardiac/cardiac_plink.assoc.logistic') %>%
  select(SNP,OR)
severe <- fread('/mnt/theta/output_data/MS_gwas_data/output/cardiac/cardiac_plink.assoc.logistic.adjusted') %>% left_join(vep) %>% left_join(or)


severe %>% ggplot(aes(x=-log10(QQ),y=-log10(UNADJ))) +
  geom_point(col='#48C095') +
  geom_smooth(method='lm', col='#27384A') +
  theme_minimal()


```


```{r q_manhattan_plot, echo=F}
severe$chr_col <- NA
severe$chr_col <- ifelse(severe$CHR %% 2 == 0, 'even','odd')
severe$chr_col <- ifelse(severe$FDR_BH <= 0.05, 's',severe$chr_col)

chr.group <- c(even = '#27384A',odd = '#48C095',s='#BC0020')

severe %>% 
  filter(is.na(UNADJ) == F) %>%
  ggplot(aes(x=CHR,y=-log10(UNADJ))) +
  geom_jitter(aes(col=chr_col)) +
  scale_color_manual(values = chr.group) +
  #geom_text(aes(label=ifelse(chr_col == 's', SYMBOL, '')), hjust=-0.3) +
  theme_classic() +
  theme(legend.position = 'none') +
  scale_x_continuous(labels = c(1:22,'X'),breaks = 1:23) +
  xlab('Chromosomes') +
  ylim(c(0,6)) +
  ylab('-log10 Pvalue')
```

```{r OR plot, echo=FALSE}
severe %>% filter(chr_col == 's') %>% distinct() %>%
  #left_join(vep_cardiac, by = 'SNP') %>% na.omit() %>%
  ggplot(aes(y=reorder(SNP,-OR),x=OR)) + 
  geom_point(col="#48C095") +
  geom_text(aes(label=SYMBOL),hjust=-0.2, vjust=0.2) +
  geom_vline(xintercept = 1,linetype='dashed',col="#BC0020") +
  ylab('SNP') +
  xlab('Odds ratio') +
  ggtitle('Estimated odds ratio (OR) of minor allele occurence', 
          subtitle = 'Red dashed line: OR = 1.0') +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust=0.5))
```




```{r echo=F}
kable(severe %>% filter(chr_col == 's') %>% distinct() %>%
        select(SNP,SYMBOL,  Consequence, UNADJ,OR) %>%
        rename(Pvalue = UNADJ)
)

```