---
title: "common variants"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressMessages(library(knitr))
suppressMessages(library(data.table))
suppressMessages(library(tidyverse))
suppressMessages(library(qvalue))

outlier <- function (x) {
  IQR <- quantile(x)[[4]] - quantile(x)[[2]]
  low <- quantile(x)[[2]] - 1.5*IQR
  high <- quantile(x)[[4]] + 1.5*IQR
  out_val <- x[x < low | x > high]
  return(out_val)
}
```

Results on 1076 unrelated individuals


```{r loading_data, echo=F}
severe <- fread('/mnt/theta/output_data/MS_gwas_data/output/fullset/multisample_0716_severe.assoc.logistic.adjusted')

vep_severe <- fread('/mnt/theta/output_data/MS_gwas_data/full_set/common_severe.tsv',header = T) %>%
  rename(SNP = Uploaded_variation) %>% select(SNP,SYMBOL,IMPACT,Consequence)

severe <- severe %>% left_join(vep_severe) %>% na.omit()

# severe %>% ggplot(aes(x=-log10(QQ),y=-log10(UNADJ))) +
#   geom_point(col='#48C095') +
#   geom_smooth(method='lm', col='#27384A') +
#   theme_minimal()

resistance <- fread('/mnt/theta/output_data/MS_gwas_data/output/fullset/multisample_0716_resist.assoc.logistic.adjusted')

vep_resist <- fread('/mnt/theta/output_data/MS_gwas_data/full_set/common_resist.tsv',header = T) %>%
  rename(SNP = Uploaded_variation) %>% select(SNP,SYMBOL,IMPACT,Consequence)

resistance <- resistance %>% left_join(vep_severe)


# resistance %>% ggplot(aes(x=-log10(QQ),y=-log10(UNADJ))) +
#   geom_point(col='#48C095') +
#   geom_smooth(method='lm', col='#27384A') +
#   theme_minimal()

```


### Severe manhattan plot

```{r q_manhattan_plot, echo=F}
severe$chr_col <- NA
severe$chr_col <- ifelse(severe$CHR %% 2 == 0, 'even','odd')
severe$chr_col <- ifelse(severe$FDR_BH <= 0.3, 's',severe$chr_col)

chr.group <- c(even = '#27384A',odd = '#48C095',s='#BC0020')

severe %>% 
  filter(is.na(UNADJ) == F) %>%
  ggplot(aes(x=CHR,y=-log10(UNADJ))) +
  geom_jitter(aes(col=chr_col)) +
  scale_color_manual(values = chr.group) +
  theme_classic() +
  theme(legend.position = 'none') +
  scale_x_continuous(labels = c(1:22,'X'),breaks = 1:23) +
  xlab('Chromosomes') +
  ylim(c(0,6)) +
  ylab('-log10 Pvalue')
```


### Top 15 severe results

```{r echo=F}
kable(head(severe,15) %>%
        select(SYMBOL, IMPACT, Consequence, SNP,UNADJ,FDR_BH,FDR_BY) %>%
        rename(Pvalue = UNADJ)
)

```


```{r manhattan_plot_resist, echo=F}
resistance$chr_col <- NA
resistance$chr_col <- ifelse(resistance$CHR %% 2 == 0, 'even','odd')
resistance$chr_col <- ifelse(resistance$FDR_BH <= 0.3, 's',resistance$chr_col)

chr.group <- c(even = '#27384A',odd = '#48C095',s='#BC0020')

resistance %>% 
  filter(is.na(UNADJ) == F) %>%
  ggplot(aes(x=CHR,y=-log10(UNADJ))) +
  geom_jitter(aes(col=chr_col)) +
  scale_color_manual(values = chr.group) +
  theme_classic() +
  theme(legend.position = 'none') +
  scale_x_continuous(labels = c(1:22,'X'),breaks = 1:23) +
  xlab('Chromosomes') +
  ylim(c(0,6)) +
  ylab('-log10 Pvalue')
```


